rfd/0096-helm-chart-revamp.md

```diff
@@ -0,0 +1,436 @@
+stays the same, this requires the loadbalancing service to remain the same.
+
+#### Teleport-specific configuration values
+
+This proposal introduces two new values for users to edit the `teleport.yaml`
+config: `auth.teleportConfig` and `proxy.teleportConfig`. The content of those
+values should be merged with the generated configuration, as described [in the
+previous section](#generating-teleportyaml).
+
+For example:
+```yaml
+auth:
+  teleportConfig:
+    auth_service:
+      authentication:
+        connector_name: "my-connector"
+proxy:
+  teleportConfig:
+    proxy_service:
+      acme:
+        enabled: true
+        email: foo@example.com
+```
+
+The following values are core values: users must set them for the chart to work
+properly. They support the happy path. Those values should not be changed by
+the proposal as it would harm backward compatibility and user experience.
+
+- `clusterName`
+- `publicAddr`
+- `chartMode`
+- `aws`
+- `gcp`
+- `enterprise`
+- `operator`
+
+The following values are used to generate Teleport's configuration, we must
+continue to support them for backward compatibility, but using
+`*.teleportConfig` should be preferred.
+
+- `kubeClusterName`
+- `authentication`
+- `authenticationType`
+- `authenticationSecondFactor`
+- `proxyListenerMode`
+- `sessionRecording`
+- `separatePostgresListener`
+- `separateMongoListener`
+- `kubePublicAddr`
+- `mongoPublicAddr`
+- `mysqlPublicAddr`
+- `postgresPublicAddr`
+- `sshPublicAddr`
+- `tunnelPublicAddr`
+- `acme`
+- `acmeEmail`
+- `acmeURI`
+- `log`
+
+#### Chart-specific configuration values
+
+Some values are used to configure the Kubernetes resources deploying Teleport.
+When specified they should apply to both auth and proxy deployments.  Those
+values are:
+
+- `podSecurityPolicy`
+- `labels`
+- `highAvailability`
+- `tls`
+- `image`
+- `enterpriseImage`
+- `affinity`
+- `annotations`
+- `extraArgs`
+- `extraEnv`
+- `extraVolumes`
+- `extraVolumeMounts`
+- `imagePullPolicy`
+- `initContainers`
+- `postStart`
+- `securityContext`
+- `priorityClassName`
+- `tolerations`
+- `probeTimeoutSeconds`
+- `teleportVersionOverride`
+- `resources`
+
+A few values will have to be treated differently:
+
+- `persistence` will only apply to the `auth` deployment
+- `service` will only apply to the `proxy` service
+- `serviceAccount.name` will apply to the `auth`, the proxy service account
+  name should be the auth one suffixed with `-proxy`
+
+Some users will need to set different values for auth and proxy pods, the
+following values should be also available under `auth` and `proxy`. Those
+specific values should take precedence over the ones at the root.
+
+- `labels`
+- `highAvailability` (except the certManager section)
+- `affinity`
+- `annotations`
+- `extraArgs`
+- `extraEnv`
+- `extraVolumes`
+- `extraVolumeMounts`
+- `initContainers`
+- `postStart`
+- `tolerations`
+- `teleportVersionOverride`
+- `resources`
+
```

Comment:
Should we also add support for auth/proxy-specific `nodeSelector` values here to make it easier for people to schedule their auth/proxy pods on different nodes if desired?

Reply:
If we support node/proxy specific affinities I don't think we need to support nodeSelector. Kubernetes made a mistake by putting nodeSelector into a GA API, now they're stuck with the field which has been technically replaced by the more powerful affinity one, but cannot be removed.